[Security Solution][Detection Engine] adds alert suppression for New Terms rule type

vitaliidm opened the pull request:
## Summary

 - addresses https://github.com/elastic/security-team/issues/8824
 - adds alert suppression for new terms rule type
 - fixes `getOpenAlerts` test function, which returned closed alerts as well
 
### UI

<img width="2294" alt="Screenshot 2024-04-02 at 12 53 26" src="https://github.com/elastic/kibana/assets/92328789/8398fba4-a06c-464b-87ef-1c5d5a18e37f">
<img width="1651" alt="Screenshot 2024-04-02 at 12 53 46" src="https://github.com/elastic/kibana/assets/92328789/971ec0da-c1d9-4c96-a4af-7cc8dfae52a4">



### Checklist
- [x] Functional changes are hidden behind a feature flag 

  Feature flag `alertSuppressionForNewTermsRuleEnabled`

- [x]  Functional changes are covered with a test plan and automated tests.

  Test plan: https://github.com/elastic/security-team/pull/9045

- [x]  Stability of new and changed tests is verified using the [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner).

Cypress ESS: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5547
Cypress Serverless: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5548

FTR ESS: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5596
FTR Serverless: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5597

- [x]  Comprehensive manual testing is done by two engineers: the PR author and one of the PR reviewers. Changes are tested in both ESS and Serverless.

- [x]  Mapping changes are accompanied by a technical design document. It can be a GitHub issue or an RFC explaining the changes. The design document is shared with and approved by the appropriate teams and individual stakeholders.

  Existing AlertSuppression schema field is used for New terms rule, the one that used for Query and IM rules. 

```yml
    alert_suppression:
      $ref: './common_attributes.schema.yaml#/components/schemas/AlertSuppression'
```
where

```yml
    AlertSuppression:
      type: object
      properties:
        group_by:
          $ref: '#/components/schemas/AlertSuppressionGroupBy'
        duration:
          $ref: '#/components/schemas/AlertSuppressionDuration'
        missing_fields_strategy:
          $ref: '#/components/schemas/AlertSuppressionMissingFieldsStrategy'
      required:
        - group_by
   ```

- [x]  Functional changes are communicated to the Docs team. A ticket or PR is opened in https://github.com/elastic/security-docs. The following information is included: any feature flags used, affected environments (Serverless, ESS, or both).

https://github.com/elastic/security-docs/issues/5030

vitaliidm committed:
[Security Solution][Detection Engine] adds alert suppression for New Terms rule type

vitaliidm commented:
/ci

vitaliidm committed:
add usint tests for converter

vitaliidm committed:
cypress tests

vitaliidm committed:
remove failures

vitaliidm commented:
/ci

vitaliidm committed:
fix tests

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm commented:
/ci

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm committed:
add ftr tests

vitaliidm committed:
split search and bulk create

vitaliidm committed:
add BE

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm commented:
/ci

vitaliidm committed:
add more fixes

vitaliidm committed:
fix tests

vitaliidm committed:
Merge branch 'de_8_14/new_terms_suppression' of https://github.com/vitaliidm/kibana into de_8_14/new_terms_suppression

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm committed:
fixes some tests

vitaliidm committed:
Merge branch 'de_8_14/new_terms_suppression' of https://github.com/vitaliidm/kibana into de_8_14/new_terms_suppression

vitaliidm commented:
/ci

vitaliidm committed:
fix lint

vitaliidm committed:
fix tests

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm commented:
/ci

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm commented:
/ci

vitaliidm committed:
fixes

vitaliidm commented:
/ci

vitaliidm committed:
refactoring/fixes

vitaliidm committed:
fix lints

vitaliidm commented:
/ci

vitaliidm committed:
add unit test

vitaliidm committed:
fix UNIT tests

vitaliidm commented:
/ci

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm commented:
/ci

vitaliidm committed:
add more tests

vitaliidm committed:
remove IM FF

vitaliidm committed:
add additional ftr exception test

vitaliidm committed:
add enrichment tests

vitaliidm commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

vitaliidm commented:
## Files by Code Owner

### elastic/obs-ux-management-team

* x-pack/plugins/rule_registry/server/utils/create_persistence_rule_type_wrapper.ts
* x-pack/plugins/rule_registry/server/utils/persistence_types.ts

### elastic/response-ops

* x-pack/plugins/alerting/server/integration_tests/__snapshots__/serverless_upgrade_and_rollback_checks.test.ts.snap
* x-pack/plugins/rule_registry/server/utils/create_persistence_rule_type_wrapper.ts
* x-pack/plugins/rule_registry/server/utils/persistence_types.ts

### elastic/security-detection-engine

* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_request_schema.test.ts
* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.gen.ts
* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/description_step/index.test.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/schema.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/use_experimental_feature_fields_transform.ts
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_creation/helpers.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/bulk_create_suppressed_alerts_in_memory.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/create_new_terms_alert_type.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/multi_terms_composite.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/types.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/wrap_suppressed_new_terms_alerts.test.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/wrap_suppressed_new_terms_alerts.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/bulk_create_suppressed_alerts_in_memory.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/bulk_create_with_suppression.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/get_is_alert_suppression_active.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/index.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/partition_missing_fields_events.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/search_after_bulk_create_suppressed_alerts.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/suppression_utils.test.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/suppression_utils.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/wrap_suppressed_alerts.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/exceptions/workflows/trial_license_complete_tier/role_based_rule_exceptions_workflows.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/serverless.config.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/esql.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/index.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/machine_learning.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/new_terms.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/new_terms_alert_suppression.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/query.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/saved_query.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threat_match.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threat_match_alert_suppression.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threshold.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threshold_alert_suppression.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/timestamps.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows_suppression_serverless_essentials.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows_supression_ess_basic.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/indicator_match_rule_suppression_ess_basic.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/new_terms_rule.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/threshold_rule_ess_basic.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/threshold_rule_serverless_essentials.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_edit/new_terms_rule.cy.ts

### elastic/security-detection-rule-management

* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_request_schema.test.ts
* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.gen.ts
* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml
* x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_definition_section.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.test.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.tsx
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.test.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/prebuilt_rules/prebuilt_rules_preview.cy.ts

### elastic/security-detections-response

* x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/alerts/get_alerts.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/alerts/get_opened_alerts.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/alerts/index.ts

### elastic/security-engineering-productivity

* x-pack/test/security_solution_cypress/config.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows_suppression_serverless_essentials.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows_supression_ess_basic.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/indicator_match_rule_suppression_ess_basic.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/new_terms_rule.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/threshold_rule_ess_basic.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/threshold_rule_serverless_essentials.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_edit/new_terms_rule.cy.ts
* x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/prebuilt_rules/prebuilt_rules_preview.cy.ts
* x-pack/test/security_solution_cypress/cypress/tasks/create_new_rule.ts
* x-pack/test/security_solution_cypress/serverless_config.ts

### elastic/security-solution

* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_request_schema.test.ts
* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.gen.ts
* x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml
* x-pack/plugins/security_solution/common/detection_engine/constants.ts
* x-pack/plugins/security_solution/common/detection_engine/utils.test.ts
* x-pack/plugins/security_solution/common/experimental_features.ts
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/description_step/index.test.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/schema.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/use_experimental_feature_fields_transform.ts
* x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_creation/helpers.ts
* x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_definition_section.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.test.tsx
* x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.tsx
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.test.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_schema/model/rule_schemas.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/bulk_create_suppressed_alerts_in_memory.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/create_new_terms_alert_type.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/multi_terms_composite.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/types.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/wrap_suppressed_new_terms_alerts.test.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/wrap_suppressed_new_terms_alerts.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/bulk_create_suppressed_alerts_in_memory.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/bulk_create_with_suppression.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/get_is_alert_suppression_active.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/index.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/partition_missing_fields_events.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/search_after_bulk_create_suppressed_alerts.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/suppression_utils.test.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/suppression_utils.ts
* x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/wrap_suppressed_alerts.ts
* x-pack/test/security_solution_api_integration/config/ess/config.base.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/exceptions/workflows/trial_license_complete_tier/role_based_rule_exceptions_workflows.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/serverless.config.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/esql.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/index.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/machine_learning.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/new_terms.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/new_terms_alert_suppression.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/query.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/saved_query.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threat_match.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threat_match_alert_suppression.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threshold.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/threshold_alert_suppression.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/timestamps.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/alerts/get_alerts.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/alerts/get_opened_alerts.ts
* x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/alerts/index.ts

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

pmuellr reviewed (approved):
LGTM

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/common_flows_supression_ess_basic.cy.ts:
```diff
@@ -0,0 +1,72 @@
+      it('can not create rule with rule execution suppression on basic license for all rules with enabled suppression', () => {
+        // Default query rule
+        cy.get(ALERT_SUPPRESSION_FIELDS_INPUT).should('be.disabled');
+        cy.get(ALERT_SUPPRESSION_FIELDS).trigger('mouseover');
```
NIT: To wrap this into a meaningful task method and the other triggers as well.

vitaliidm replied:
wrapped in a separate utility

MadameSheema reviewed (approved):
security-engineering-productivity changes LGTM, even so, please take into consideration the following:

- As the FF has been enabled in `x-pack/test/security_solution_cypress/config.ts` all the tests executed for both ESS and Serverless on PRs will be affected by that change, not only the ones were are specifically set at the top of the spec file.

- On MKI environments, we don't have the ability to enable FF through the automation, once [the changes asked regarding tagging](https://github.com/elastic/kibana/pull/179737#discussion_r1546612940) are implemented, all the serverless tests that have the feature flag will fail on the periodic pipeline. cc @yctercero 

- I saw you are doing changes as well in archives living in `x-pack/test/functional/es_archives/security_solution `, [now we are able to reuse archives](https://github.com/elastic/kibana/pull/178724) on that location in our Cypress tests as well. Letting you know in case that this new functionality might facilitate your work.

vitaliidm committed:
CR feedback

vitaliidm committed:
lint errors

maximpn reviewed (approved):
Rule Management Area LGTM 👍 

I checked the PR locally in ESS and Serverless for one field alert suppression and haven't found any problems.

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm committed:
remove unnecessary file

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/bulk_create_suppressed_alerts_in_memory.ts:
```diff
@@ -0,0 +1,112 @@
+  experimentalFeatures: ExperimentalFeatures | undefined;
+}
+/**
+ * bulk create and suppress alerts in memory,
```
Nit: these comments don't provide much more information than the name does. I would mention that this is explicitly for new terms alerts, or at least deals with `EventsAndTerms[]` instead of normal alerts.

The note about missing fields logic is nice, but it would be nice to discuss it more explicitly, i.e. "if {@param alertSuppression.missingFieldsStrategy} is X, alerts missing the suppression field(s) will not be suppressed."

But: typing out the above, this doesn't feel like the right place to document that behavior. I would capture it in documentation, and ideally a unit test for this function, otherwise an integration test of some kind.

vitaliidm replied:
I added a bit more details to description, mentioned difference with common utility.

There multiple FTR tests that cover missing fields logic: https://github.com/elastic/kibana/blob/ff8a41fdf43e748c214b6b1e535aaac4bc28669a/x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/new_terms_alert_suppression.ts#L874-L981

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/get_is_alert_suppression_active.ts:
```diff
@@ -0,0 +1,45 @@
+ * - feature flag should not be disabled
+ * - license should be platinum
+ */
+export const getIsAlertSuppressionActive = async ({
```
Note: this exists as [`isAlertSuppressionActive`](https://github.com/elastic/kibana/pull/176422/files#diff-f3b55ecbc2dfc70ebdb1af1ac6591b982d63259d9ae1271320ad4eca7bf8647aR20) on the EQL suppression branch. We should probably reconcile those once this is merged.

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/partition_missing_fields_events.ts:
```diff
@@ -15,16 +16,20 @@ import type { SignalSourceHit } from '../types';
+>(
+  events: T[],
+  suppressedBy: string[] = [],
+  fieldsPath: ['event'] | [] = []
```
This parameter is pretty confusing, and a lack of tests/documentation doesn't help. Why is this an array (as opposed to a dotted string, etc)? Why is `['event']` the only valid value?

vitaliidm replied:
added comment and test.
event only valid value for now as fields property can be located at `event` key for new terms rule.
Which operates with object of event and new term, instead of just event as the rest of the rules

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/search_after_bulk_create_suppressed_alerts.ts:
```diff
@@ -55,62 +50,20 @@ export const searchAfterAndBulkCreateSuppressedAlerts = async (
-      alertWithSuppression,
-      ruleExecutionLogger,
-      wrappedDocs,
+    return bulkCreateSuppressedAlertsInMemory({
```
💯

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/suppression_utils.ts:
```diff
@@ -0,0 +1,86 @@
+};
+
+/**
+ * retrieves suppression values from source fields based on alert suppression configuration
```
```suggestion
 * returns an array of {@link SuppressionTerm}s by retrieving the appropriate field values based on the provided {@param alertSuppression}
```

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/suppression_utils.test.ts:
```diff
@@ -0,0 +1,102 @@
+      })
+    ).toEqual([{ field: 'host.name', value: 'localhost-1' }]);
+  });
+  it('should return sorted suppression terms array', () => {
```
An example that returns multiple terms would be nice! And maybe a case where there's suppression but nothing in `fields`? More edge cases.

vitaliidm replied:
nice suggestion! Added more test cases

rylnd commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/wrap_suppressed_alerts.ts:
```diff
@@ -102,22 +83,19 @@ export const wrapSuppressedAlerts = ({
-        [ALERT_SUPPRESSION_END]: suppressionTime,
-        [ALERT_SUPPRESSION_DOCS_COUNT]: 0,
-        [ALERT_INSTANCE_ID]: instanceId,
+        ...getSuppressionAlertFields({
```
👍

rylnd reviewed (approved):
I gave a first pass here and everything looks GREAT. Pretty straightforward overall, with some reasonable refactorings that didn't explode the diff too much.

I had a few nits, but they were mainly around documentation/posterity stuff.

I have NOT reviewed the integration/cypress tests yet. I plan to go through those alongside the EQL suppression branch to compare/contrast coverage and approach. If I have any suggestions for adding coverage, we can address those after this gets merged.

Between the plentiful tests, the green build, and the feature flag, I'm approving this as is. I did check this out locally, tried a few different suppression scenarios, tried to break the rule creation form, etc. Everything worked for me!

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_definition_section.tsx:
```diff
@@ -645,6 +645,28 @@ const prepareDefinitionSectionListItems = (
     });
   }
 
+  if ('new_terms_fields' in rule && rule.new_terms_fields && rule.new_terms_fields.length > 0) {
```
Yes, changes in schema still needed. This part is responsible for rule details page only.
Schema part is responsible for details view on rule create form, step preview

<img width="1453" alt="Screenshot 2024-04-08 at 10 20 32" src="https://github.com/elastic/kibana/assets/92328789/505fca72-e6fe-4051-9021-0c391616d9de">

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.test.tsx:
```diff
@@ -7,8 +7,28 @@
+  .mockReturnValue(false);
 
 describe('useAlertSuppression', () => {
+  it('should return isSuppressionEnabled false if rule Type exists in SUPPRESSIBLE_ALERT_RULES and Feature Flag is disabled', () => {
```
a lot of other tests have `should`, which is used widely across app.
What the criteria to use/not use `should` in your opinion?
Maybe based on it, we can draw some guideline on its usage?

vitaliidm committed:
CR feedback

vitaliidm commented on x-pack/test/functional/es_archives/security_solution/ecs_compliant/mappings.json:
```diff
@@ -67,6 +67,9 @@
             "name": {
               "type": "keyword"
             },
+            "ip": {
```
Some of the ftr tests are using `host.ip` property, that's why it added to mappings here

vitaliidm committed:
CR: feedback

vitaliidm committed:
CR: add more test cases

vitaliidm committed:
CR feedback

vitaliidm commented:
> I gave a first pass here and everything looks GREAT. Pretty straightforward overall, with some reasonable refactorings that didn't explode the diff too much.
I had a few nits, but they were mainly around documentation/posterity stuff.
I have NOT reviewed the integration/cypress tests yet. I plan to go through those alongside the EQL suppression branch to compare/contrast coverage and approach. If I have any suggestions for adding coverage, we can address those after this gets merged.
Between the plentiful tests, the green build, and the feature flag, I'm approving this as is. I did check this out locally, tried a few different suppression scenarios, tried to break the rule creation form, etc. Everything worked for me!

Thanks for review and feedback, @rylnd 
I have addressed comments and once it merged we can go ahead with updating EQL PR and resolve possible conflicts

vitaliidm committed:
fix jest test

vitaliidm committed:
Merge branch 'main' into de_8_14/new_terms_suppression

vitaliidm committed:
fix FTR test
